inst/tests/tests.Rraw

```diff
@@ -13955,6 +13956,18 @@ DT = data.table(
   V9 = c(0.2, -0.1, 1.2, -0.5, 1.4, 1, 0.2, 0.7, 0.4, 1.6),
   V10 = c(0.8, 0.7, -1.2, -0.9, -0.6, 0.4, -2.3, 2.2, 0.5, -1.4)
 )
+test(1971.01, DT[ , lapply(.SD, sum), .SDcols = patterns('^V', cols="V7")], error="user should not provide cols argument to patterns, when specifying the columns for melt or .SDcols; in this context, non-standard evaluation is used internally to set cols to all data table column names, so please fix by removing cols argument")
+not_patterns <- function()c("V4","V8")
+test(1971.02, DT[ , lapply(.SD, sum), .SDcols = not_patterns()], data.table(V4=3.4, V8=-0.4))
+not_patterns <- function(cols)cols
+test(1971.03, DT[ , lapply(.SD, sum), .SDcols = not_patterns()], error='argument "cols" is missing, with no default')
+test(1971.04, DT[ , lapply(.SD, sum), .SDcols = not_patterns(c("V4","V8"))], data.table(V4=3.4, V8=-0.4))
+patterns <- function(cols)list(cols[c(6,10)])#with cols arg
```

Comment:
ooh this is funky. this feels like something users should not rely on working... I would probably drop this test. WDYT?

Reply:
yes I agree this is funky. Did you see my comment about this above? https://github.com/Rdatatable/data.table/pull/5115#pullrequestreview-1904158739 So to move forward I would suggest either
* keep current behavior, NSE in .SDcols only used with a function named patterns(), with user-defined patterns() un-documented and un-tested.
* modify current behavior so that NSE is used in .SDcols with any user-defined function (whether named patterns or not), and add new docs/tests. I believe this option should be preferable for consistency of how functions are evaluated in the context of .SDcols, and how eval_with_cols is used here and in melt.

@MichaelChirico can you please tell me which option you would prefer? After there is consensus, I will update the PR.

Reply:
I lean towards the former. We really are trying to minimize our use of NSE across the API, adding new places where it works outside `i=`, `j=`, `by=` is not recommended IMO. See e.g. #6035 where even very basic NSE is rejected.

Reply:
ok I removed the docs and tests which mention user-defined patterns